subprojects/tooling-api/src/main/java/org/gradle/tooling/PhasedBuildActionExecuter.java

```diff
@@ -0,0 +1,142 @@
+         * <p>Models contributed by project plugins won't be available at this point.
+         *
+         * <p>If the operation fails, build will fail with the appropriate exception. Handler won't be notified in case of failure.
+         *
+         * @param buildAction The action to run in the specified build phase.
+         * @param handler The handler to supply the result of the given action to.
+         * @param <T> The returning type of the action.
+         * @return The builder.
+         * @throws IllegalArgumentException If an action has already been added to this build phase. Multiple actions per phase are not supported yet.
+         */
+        <T> Builder projectsLoaded(BuildAction<T> buildAction, PhasedResultHandler<? super T> handler) throws IllegalArgumentException;
+
+        /**
+         * Executes the given action after projects are evaluated and sends its result to the given result handler.
+         *
+         * <p>If the operation fails, build will fail with the appropriate exception. Handler won't be notified in case of failure.
+         *
+         * @param buildAction The action to run in the specified build phase.
+         * @param handler The handler to supply the result of the given action to.
+         * @param <T> The returning type of the action.
+         * @return The builder.
+         * @throws IllegalArgumentException If an action has already been added to this build phase. Multiple actions per phase are not supported yet.
+         */
+        <T> Builder projectsEvaluated(BuildAction<T> buildAction, PhasedResultHandler<? super T> handler) throws IllegalArgumentException;
+
+        /**
+         * Executes the given action after tasks are run and sends its result to the given result handler.
+         *
+         * <p>If the operation fails, build will fail with the appropriate exception. Handler won't be notified in case of failure.
+         *
+         * @param buildAction The action to run in the specified build phase.
+         * @param handler The handler to supply the result of the given action to.
+         * @param <T> The returning type of the action.
+         * @return The builder.
+         * @throws IllegalArgumentException If an action has already been added to this build phase. Multiple actions per phase are not supported yet.
+         */
+        <T> Builder buildFinished(BuildAction<T> buildAction, PhasedResultHandler<? super T> handler) throws IllegalArgumentException;
+
+        /**
+         * Builds the executer from the added actions.
+         *
+         * @return The executer.
+         */
+        PhasedBuildActionExecuter build();
+    }
+
+    /**
+     * Specifies the tasks to execute before executing the BuildFinishedAction and after the ProjectsEvaluatedAction.
+     *
+     * The graph task can be changed in model builders invoked. If not configured, null, or an empty array is passed, then no tasks will be executed unless one of the model builders configures it.
```

Comment:
❌ This is not true. Tasks will always be run. I think we should remove this paragraph and instead explain task execution later when we add the "upToPhase" concept. Also let's not call out the possibility of mutating the start parameters in the model builder.